File: app/src/main/java/net/emilla/math/Maths.java
package net.emilla.math;

import android.os.Build;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.text.DecimalFormat;

public enum Maths {;
    public static String prettyNumber(BigDecimal n) {
        // todo: configurable sig digs.
        return new DecimalFormat("#.######").format(n);
    }

    public static int exactInt(BigInteger n) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return n.intValueExact();
        }

        if (n.bitLength() <= 31) {
            return n.intValue();
        }

        throw new ArithmeticException("BigInteger out of int range");
    }

    public static NumberFormatException malformedExpression() {
        return new NumberFormatException("Malformed expression");
    }

    public static BigInteger stupidFactorial(BigInteger n) {
        if (n.compareTo(BigInteger.ZERO) < 0) {
            throw new ArithmeticException("Can't compute factorial of negative value");
        }

        var factorial = BigInteger.ONE;
        for (var m = BigInteger.valueOf(2); m.compareTo(n) <= 0; m = m.add(BigInteger.ONE)) {
            factorial = factorial.multiply(m);
        }

        return factorial;
    }
}

File: app/src/main/java/net/emilla/math/Maths.kt
@file:JvmName("Maths")

package net.emilla.math

import androidx.annotation.StringRes
import net.emilla.R
import net.emilla.exception.EmillaException
import java.text.DecimalFormat

// todo: configurable sig digs.
fun prettyNumber(n: Double): String = DecimalFormat("#.######").format(n)

fun tryParseLong(num: String, @StringRes errorTitle: Int) = try {
    num.toLong()
} catch (_: NumberFormatException) {
    throw malformedExpression(errorTitle)
}

fun tryParseDouble(num: String, @StringRes errorTitle: Int) = try {
    num.toDouble()
} catch (_: NumberFormatException) {
    throw malformedExpression(errorTitle)
}

fun malformedExpression(@StringRes errorTitle: Int): EmillaException {
    return EmillaException(errorTitle, R.string.error_calc_malformed_expression)
}

fun undefined(@StringRes errorTitle: Int): EmillaException {
    return EmillaException(errorTitle, R.string.error_calc_undefined)
}

fun Double.factorial(): Double {
    return toLong().factorial().toDouble()
    // TODO: this discards the fractional part. Find the proper way to compute factorial of
    //  fractional numbers
}

fun Long.factorial(): Long {
    if (this < 0L) throw ArithmeticException()
    if (this < 2L) return 1L

    var fact = 2L
    for (i in 3L..this) fact *= i

    return fact
}
